src/MySqlConnector/MySqlClient/Results/ResultSet.cs

```diff
@@ -0,0 +1,373 @@
+				if (firstByte == OkPayload.Signature)
+				{
+					var ok = OkPayload.Create(payload);
+					RecordsAffected = ok.AffectedRowCount;
+					LastInsertId = ok.LastInsertId;
+					ColumnDefinitions = null;
+					State = ok.ServerStatus.HasFlag(ServerStatus.MoreResultsExist)
+						? ResultSetState.HasMoreData
+						: ResultSetState.NoMoreData;
+					if (State == ResultSetState.NoMoreData)
+						break;
+				}
+				else if (firstByte == 0xFB)
+				{
+					throw new NotSupportedException("Don't support LOCAL_INFILE_Request");
+				}
+				else
+				{
+					var reader = new ByteArrayReader(payload.ArraySegment);
+					var columnCount = (int) reader.ReadLengthEncodedInteger();
+					ColumnDefinitions = new ColumnDefinitionPayload[columnCount];
+					m_dataOffsets = new int[columnCount];
+					m_dataLengths = new int[columnCount];
+
+					for (var column = 0; column < ColumnDefinitions.Length; column++)
+					{
+						payload = await Session.ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+						ColumnDefinitions[column] = ColumnDefinitionPayload.Create(payload);
+					}
+
+					payload = await Session.ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+					EofPayload.Create(payload);
+
+					LastInsertId = -1;
+					State = ResultSetState.ReadResultSetHeader;
+					break;
+				}
+			}
+			BufferState = State;
+			//await BufferEntireAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+			return this;
+		}
+
+		public async Task BufferEntireAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			while (BufferState == ResultSetState.ReadingRows || BufferState == ResultSetState.ReadResultSetHeader)
+				await BufferReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+		}
+
+		public async Task ReadEntireAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			while (State == ResultSetState.ReadingRows || State == ResultSetState.ReadResultSetHeader)
+				await ReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+		}
+
+		public bool Read()
+		{
+			return ReadAsync(IOBehavior.Synchronous, CancellationToken.None).GetAwaiter().GetResult();
+		}
+
+		public Task<bool> ReadAsync(CancellationToken cancellationToken) =>
+			ReadAsync(Command.Connection.AsyncIOBehavior, cancellationToken);
+
+		public async Task<bool> ReadAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			m_row?.ClearData();
+			m_row = m_readBuffer.Count > 0
+				? m_readBuffer.Dequeue()
+				: await DoReadAsync(ioBehavior, m_row, cancellationToken).ConfigureAwait(false);
+
+			if (m_row == null)
+			{
+				State = BufferState;
+				return false;
+			}
+			State = ResultSetState.ReadingRows;
+			return true;
+		}
+
+		public async Task<Row> BufferReadAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			m_rowBuffered?.BufferData();
+			m_rowBuffered = await DoReadAsync(ioBehavior, null, cancellationToken).ConfigureAwait(false);
```

Comment:
Won't `m_rowBuffered` already be set by `DoReadAsync`? Unless it returns `null`, in which case `null` will be enqueued in `m_readBuffer`? Could that cause problems later?

It's confusing, and potentially error-prone, to have multiple functions responsible for setting `m_rowBuffered`. (Still trying to ensure I understand the logic properly before suggesting an alternative.)